run-webkit-tests --print-expectations -v lists all considered TextExpectations lines

squelart opened the pull request:
#### 70c0d8553e42c0cb5209bab2b8ffd6f813e42d0c
<pre>
run-webkit-tests --print-expectations -v lists all considered TextExpectations lines
<a href="https://bugs.webkit.org/show_bug.cgi?id=269895">https://bugs.webkit.org/show_bug.cgi?id=269895</a>
<a href="https://rdar.apple.com/123424189">rdar://123424189</a>

Reviewed by Sam Sneddon.

Instead of only showing the one line that was used to set a test&apos;s expectations,
we can now show all the lines that were also considered before and/or after.

A unit test was added to verify that the non-verbose output matches exactly with
the expected output, to avoid issues with buildbots that crudely diff this
output between revisions.

* Tools/Scripts/webkitpy/layout_tests/controllers/manager.py:
(Manager._print_expectation_line_for_test):
(Manager._print_expectations_for_subset):
* Tools/Scripts/webkitpy/layout_tests/controllers/manager_unittest.py:
(ManagerTest._get_manager):
(ManagerTest.test_look_for_new_crash_logs):
(ManagerTest):
(ManagerTest.test_print_expectations_for_subset):
(ManagerTest.test_print_expectations_for_subset.get_test_names):
(ManagerTest.test_print_expectations_for_subset.get_tests):
(ManagerTest.test_print_expectations_for_subset.get_expectations):
(get_printed_expectations):
(parse_exp):
(ManagerTest.test_look_for_new_crash_logs.get_manager): Deleted.
* Tools/Scripts/webkitpy/layout_tests/models/test_expectations.py:
(TestExpectationsModel.__init__):
(TestExpectationsModel.get_expectation_lines):
(TestExpectationsModel.add_expectation_line):
* Tools/Scripts/webkitpy/layout_tests/models/test_expectations_unittest.py:
* Tools/Scripts/webkitpy/layout_tests/run_webkit_tests.py:
(parse_args):

Canonical link: <a href="https://commits.webkit.org/283176@main">https://commits.webkit.org/283176@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c49b6c9733d4a80e7b686acebcf6a0730a03a889

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/65371 "3 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/44740 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/17985 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/69395 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/15977 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/67489 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/52600 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/16259 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/52494 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/11051 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/68437 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/41404 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/56578 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/33118 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/64883 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/38076 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/13954 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/14854 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/59912 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/14293 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/71100 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/9323 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/13750 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/59818 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/9355 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/56640 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/60093 "Found 3 new API test failures: /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/accessible/list-markers, /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/component/scroll-to, /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/text/replaced-objects (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14427 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/7700 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/1357 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/40551 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/41627 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/42808 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/41371 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

gsnedders commented on Tools/Scripts/webkitpy/layout_tests/models/test_expectations.py:
```diff
@@ -715,6 +720,10 @@ def add_expectation_line(self, expectation_line, in_skipped=False):
         for test in expectation_line.matching_tests:
+            if test not in self._test_to_expectation_lines:
+                self._test_to_expectation_lines[test] = []
+            self._test_to_expectation_lines[test].append(expectation_line)
```
Performance is unlikely to be too critical here, but we can avoid having to access the item on the dict multiple times:

```suggestion
            self._test_to_expectation_lines.setdefault(test, []).append(expectation_line)
```

squelart replied:
Didn't know about `setdefault`, neat, thanks!

gsnedders commented on Tools/Scripts/webkitpy/layout_tests/controllers/manager.py:
```diff
@@ -726,14 +726,34 @@ def _worker_number(worker_name):
+                    before_main_line = False
+                elif before_main_line:
+                    decision = "v  "
+                else:
+                    decision = "^  "
```
Is this case ever hit? I thought it was _always_ the last line. Some tests would probably help clarify the difference between `get_expectation_line` and `get_expectation_lines` if that's not the case.

Because if it's the last line, then we effectively have the same behaviour for `lines[-1]` regardless of whether `verbose` is set, right? The only difference is `decision` being `"-> "` instead of `""`.

squelart replied:
Yes, it can happen. See `add_expectation_lines` in test_expectations.py, there's a test with `not in_skipped and self._already_seen_better_match()` that can discard later lines for the same file.
And it happens in practice, see the example in https://bugs.webkit.org/show_bug.cgi?id=269895  -- maybe it's a sign that one of those two lines is redundant? But in any case I think it's useful information.

I didn't see tests for `get_expectation_line`, that's why I lazily didn't bother testing `get_expectation_lines` 😅 But yeah, it could clarify what's supposed to happen.

I don't quite understand your last paragraph with the `lines[-1]`.
In non-verbose mode, we'll always use a literal "" (at line 732).
With verbose, we'll always have a 4-character decision containing either "v", "->", or "^".

squelart replied:
I've added some tests for the basic functionality, the last one shows how the "main" expectation line may not necessarily be the very last one (in that case, a later override of the whole directory, but with the same expected result.)

gsnedders commented on Tools/Scripts/webkitpy/layout_tests/controllers/manager.py:
```diff
@@ -726,14 +726,34 @@ def _worker_number(worker_name):
+                                           line.expected_behavior,
+                                           self._expectations[device_type].readable_filename_and_line_number(line),
+                                           line.original_string or ''))
+                test_path = ' ' * len(test_path)
```
I'm confused why we're doing this? It isn't immediately apparent _why_ we're replacing the test path with whitespace.

squelart replied:
That's to avoid repeating the same file over multiple lines.
Without replacement:
```
path v  skip
path -> pass
path ^  pass
```
With replacement:
```
path v  skip
     -> pass
     ^  pass
```
Should I add a comment, or do something else to make it clearer?

squelart replied:
I've decided to remove this little bit of cleverness. The output will be a bit busier with repeated filenames, but it's still readable, and more importantly it would survive grep-ing for subsets without losing context.

gsnedders reviewed (commented):
[non-reviewer]

squelart committed:
run-webkit-tests --print-expectations -v lists all considered TextExpectations lines
https://bugs.webkit.org/show_bug.cgi?id=269895
rdar://123424189

Reviewed by Sam Sneddon.

Instead of only showing the one line that was used to set a test's expectations,
we can now show all the lines that were also considered before and/or after.

A unit test was added to verify that the non-verbose output matches exactly with
the expected output, to avoid issues with buildbots that crudely diff this
output between revisions.

* Tools/Scripts/webkitpy/layout_tests/controllers/manager.py:
(Manager._print_expectation_line_for_test):
(Manager._print_expectations_for_subset):
* Tools/Scripts/webkitpy/layout_tests/controllers/manager_unittest.py:
(ManagerTest._get_manager):
(ManagerTest.test_look_for_new_crash_logs):
(ManagerTest):
(ManagerTest.test_print_expectations_for_subset):
(ManagerTest.test_print_expectations_for_subset.get_test_names):
(ManagerTest.test_print_expectations_for_subset.get_tests):
(ManagerTest.test_print_expectations_for_subset.get_expectations):
(get_printed_expectations):
(parse_exp):
(ManagerTest.test_look_for_new_crash_logs.get_manager): Deleted.
* Tools/Scripts/webkitpy/layout_tests/models/test_expectations.py:
(TestExpectationsModel.__init__):
(TestExpectationsModel.get_expectation_lines):
(TestExpectationsModel.add_expectation_line):
* Tools/Scripts/webkitpy/layout_tests/models/test_expectations_unittest.py:
* Tools/Scripts/webkitpy/layout_tests/run_webkit_tests.py:
(parse_args):

Canonical link: https://commits.webkit.org/283176@main
